### DISCUSSION TITLE
Temperature (and time) dependence for materials

### DISCUSSION THREAD

**Discussion by Jake-HW-Miles** [idaholab/moose#27348] [Q&A General]
Hi there, just getting used to using moose and I want to have material properties that change depending on the temperature. I have tried a few different things, like using parsed functions to feed into parsedmaterials or the built in function capabilities for something like thermal conductivity. 

There must be a simplier way to do this, I have provided the current state of my material block which runs but doesnt converge.... 

Any help is appreciated! (I have had to change the functions and names of blocks etc, however the structure is the same as I use)

[Materials] 
    [fluid_cp_dT_parsed]
        type = ParsedMaterial
        block = 'fluid'
    
        expression = '(1 / ( temperature^(2) ))'
        property_name = specific_heat_dT
        coupled_variables = 'temperature'
    []

    [fluid_cp_parsed]
        type = ParsedMaterial
        block = 'fluid'
    
        expression = '1'
        property_name = specific_heat
        coupled_variables = 'temperature'
    []

    [fluid_thrm_cond_dT_parsed]
        type = ParsedMaterial
        block = 'fluid'
    
        expression = '(1 / ( temperature^(2) ))'
        property_name = thermal_conductivity_dT
        coupled_variables = 'temperature'
    []

    [fluid_thrm_cond_parsed]
        type = ParsedMaterial
        block = 'fluid'
        expression = '1'
        property_name = thermal_conductivity
        coupled_variables = 'temperature'
    []

    [fluid_density_parsed]
        type = ParsedMaterial
        block = 'fluid'
        expression = '(1 / ( temperature^(2) ))'
        property_name = density
        coupled_variables = 'temperature'
    []
[]

**Comment by ngrilli** [comment]
@Jake-HW-Miles I would normally implement time dependence of a material property
by passing a function of time to the material object, see for instance this one where we implement
a time depend coefficient for creep:

https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/CrystalPlasticityDislocationUpdate.C

"creep_ao_function" is the relevant one. You can see an example input file of our recent paper on creep:

https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/Paper_CP_PFF/polycrystal_right_elastoplastic_variableGc.i

see the function "creep_rate_prefactor".

Temperature dependence is in the same material object above. 
You can see we have made the crystal plasticity laws temperature dependent
by couping "temperature".

Best Regards,
Nicolò

**Comment by GiudGiud** [comment]
Hello

First you should make sure it converges with constants everywhere. Then introduce dependencies on nonlinear variables one at a time. 

If you can use AD (for example with the ADParsedMaterial) you should try to do that. It will make sure the derivatives of the quantities with regards to the nonlinear variables are well captured

Guillaume
